### ISSUE TITLE
How can I disable AVX?

### ISSUE THREAD

**Issue by themightyoarfish** [PointCloudLibrary/pcl#4554]
# tldr

How can I disable AVX for a client program?

# long story

I have the problem that my code produces `Eigen::Transform` objects with `EIGEN_DONT_VECTORIZE` enabled, which I assume leads to objects not being aligned for CPU intrinsics. In any event, the code produces transforms which are not 32-byte aligned as required [_mm256_load_pd](https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/intrinsics/intrinsics-for-intel-advanced-vector-extensions/intrinsics-for-load-and-store-operations-1/mm256-load-pd.html).

Hence my calls to `pcl::transformPointCloud()` fails because it takes this code path:
https://github.com/PointCloudLibrary/pcl/blob/b25612f8ff0318724f3a6caae4e9e753adc2b74c/common/include/pcl/common/impl/transforms.hpp#L194

I don't understand why this happens, considering I am not passing any options such as `-march=native` to the compiler, specifically to avoid AVX (I've had nothing but unreproducible problems with intrinsics use with Eigen and PCL). I'm getting such problems here
```
* thread #9: tid = 29275, 0x000000000048c9f6 autocrane-core_main`pcl::detail::Transformer<double>::Transformer(Eigen::Matrix<double, 4, 4, 0, 4, 4> const&) [inlined] _mm256_load_pd(__P=0x00007fffb4b19370) + 4 at avxintrin.
h:836, name = 'autocrane-core_', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0x000000000048c9f6 autocrane-core_main`pcl::detail::Transformer<double>::Transformer(Eigen::Matrix<double, 4, 4, 0, 4, 4> const&) [inlined] _mm256_load_pd(__P=0x00007fffb4b19370) + 4 at avxintrin.h:836
    frame #1: 0x000000000048c9f2 autocrane-core_main`pcl::detail::Transformer<double>::Transformer(this=0x00007fffb4b19100, tf=0x00007fffb4b19370) + 100 at transforms.hpp:194
    frame #2: 0x0000000000490036 autocrane-core_main`void pcl::transformPointCloud<pcl::PointXYZI, double>(cloud_in=0x000000000157d740, cloud_out=0x00007fffb4b192e0, transform=0x00007fffb4b19370, copy_all_fields=true) + 43
9 at transforms.hpp:241
```
presumably because the affine transform matrix does not have 32 byte alignment (it works if I align it manually before calling `transformPointCloud()` with it.

It seems that since I can see the location in the debugger that the avx usage is not baked into the compiled library, so I should be able to avoid it. I am compiling PCL 1.11.1 from source, also passing `-DEIGEN_DONT_VECTORIZE` via `CXXFLAGS` env var.

## what does not help
I compiled with `-mno-avx` and also tried to `#undef __AVX__` before the PCL includes in the file, but still it takes the avx code path.

**Update**: trying `-mno-avx` again in another place of the compiler invocation _does_ work.

**Comment by mvieth** [comment]
As you assumed correctly, the transforms code ([`transforms.hpp`](https://github.com/PointCloudLibrary/pcl/blob/master/common/include/pcl/common/impl/transforms.hpp)) is not in the compiled library, instead the header file is included completely in your project and compiled there.
My first advice would be for you to check again the flags that your project is compiled with. Maybe `-march=native` is set by default? Are you sure `-mno-avx` overrides all other flags? Maybe try to also disable SSE2? I really don't see a reason why the compiler would disregard `#if !defined(__AVX__)` and the other preprocessor directives.

**Comment by themightyoarfish** [comment]
Actually I found that `-march=native` is in fact erroneously included, so this is the cause.

The question for me is: Can I disable AVX for just PCL via preprocessor directives? Or do I really need to compile this one compilation unit with a flag negating `-march=native`?

**Comment by themightyoarfish** [comment]
I assume that between `transform.h` and `transforms.hpp`, the `Eigen` includes bring int the `__AVX__` macro again somehow, but I don't have an understanding what `-march=native` does exactly in terms of preprocessor defines.

**Comment by mvieth** [comment]
> I assume that between transform.h and transforms.hpp, the Eigen includes bring int the __AVX__ macro again somehow

Seems rather unlikely to me. Can you put tests into transforms.h and transforms.hpp to see where `__AVX__` is defined and where not?

> I don't have an understanding what -march=native does exactly in terms of preprocessor defines.

[This stackoverflow question might help you](https://stackoverflow.com/questions/5470257/how-to-see-which-flags-march-native-will-activate).

**Comment by themightyoarfish** [comment]
I checked this assumption and I can't find any included code that would `#define __AVX__`, neither in PCL nor Eigen. 

Interesting observation: if I put this here in `transforms.hpp` in the spot where the error occurs, the code compiles fine
```
191   Transformer(const Eigen::Matrix4d& tf)
192   {
193 #error "wat"
194     for (std::size_t i = 0; i < 4; ++i)
195       c[i] = _mm256_load_pd (tf.col (i).data ());
196   }
```
so this code is not actually compiled. Nevertheless, the debugger complains that the error occurs in that exact location. I don't have any other versions of PCL installed (and am using specific version via cmake anyway), so I cannot explain this.

Any ideas?

**Comment by themightyoarfish** [comment]
Could it be because the same header is included with AVX in other translation units and during linking, that version is used? Not sure how the linker chooses implementations in such a case.

**Comment by mvieth** [comment]
The problem happens when you directly call `transformPointCloud`, right?

**Comment by tkircher** [comment]
I've also had endless problems with Eigen and PCL.

**Comment by mvieth** [comment]
> I've also had endless problems with Eigen and PCL.

Respectfully, stating that here is unhelpful. For what it is worth, there are not many issues here on github regarding Eigen, and I personally never had a problem with that. If you have an idea or a possible solution for this specific problem, you are very welcome to contribute here.

**Comment by tkircher** [comment]
@mvieth To some degree, PCL is responsible for supporting popular applications that use the library. So while there may not be any issues mentioned here specifically, there definitely are elsewhere downstream. Like the OP mentioned, there are architecture-specific alignment issues that have caused no shortage of problems with libraries and applications that use Eigen (AliceVision has Eigen alignment disabled by default), so they have helpfully included some explanations in their troubleshooting guide:

https://eigen.tuxfamily.org/dox/TopicPitfalls.html

Since PCL uses Eigen, this project is, to some degree, also responsible for handling this.

**Comment by themightyoarfish** [comment]
Can't say I agree with this, but it's comforting to know I'm not alone with these issues. I've been observing Eigen's guide and still always had alignment issues cropping up on some system until I disabled all vectorization. Which is why I was very surprised to have this problem again now. I could try to make a minimally reproducible example, but this invariably fails to reproduce these issues in my experience.

**Comment by mvieth** [comment]
What do you think about replacing the aligned loads (`*_load_*`) with unaligned loads (`*_loadu_*`) in the constructor? @themightyoarfish could you try if this would work for you? The constructor is not time critical, so the light performance penalties should be acceptable.

**Comment by themightyoarfish** [comment]
Thanks for all the ideas. I have recompiled a couple of times, verifying that the AVX-enabled branch is in fact compiled in when I don't take measures against it (so the original case). But now it doesn't crash, despite the fact that nothing in the code changed compare to when I first saw this problem.

I wish this was the first time I saw such unpredictable behaviour.
